[Experimental] CircleCI

aharpervc opened the pull request:
1. this is an experiment
2. supersedes https://github.com/aharpervc/tiny_tds/pull/3
3. see results here: https://app.circleci.com/pipelines/github/veracross/tiny_tds?branch=circleci

aharpervc committed:
wip: trying out circleci w/ job matrix parameters

aharpervc committed:
wip: restore sleep for database

aharpervc committed:
wip: attempt cross build

aharpervc committed:
wip: remote docker

aharpervc committed:
temp: disable linux job

aharpervc committed:
wip: remove remote docker

aharpervc committed:
wip: cross build job

aharpervc committed:
wip: add ruby version parameter

aharpervc committed:
wip: remote docker for cross build

aharpervc committed:
wip: only one version of ruby for now, for speed

aharpervc committed:
wip: cross build ports instead

aharpervc committed:
wip: remove cross build

aharpervc committed:
wip: build natively on windows

aharpervc committed:
wip: ridk install

aharpervc committed:
wip: install mingw separately (silently)

aharpervc committed:
fixup: I meant msys2

aharpervc committed:
wip: what happens if we skip all these steps

aharpervc committed:
wip: try and figure out this perl module stuff

aharpervc committed:
wip: where's perl?

aharpervc committed:
wip: order of operations

aharpervc committed:
wip: forgot to enable

aharpervc committed:
wip: extend timeout

aharpervc committed:
wip: split ports to separate steps

aharpervc committed:
wip: more time

aharpervc committed:
temp: logging to see where is the problem

aharpervc committed:
wip
- re-enable some diagnostics
- update pacman before installing ridk

aharpervc committed:
wip: use orb to install ruby

aharpervc committed:
wip: replace orb with choco install ruby

aharpervc committed:
wip: replace choco with scoop for ruby

aharpervc committed:
wip: install scoop manually first

aharpervc committed:
wip: update scoop after install

aharpervc committed:
wip: add scoop bucket for versioned ruby manifests

aharpervc committed:
wip: activate the current ruby version

aharpervc committed:
wip: scoop reset each time, sigh

aharpervc committed:
wip: manually install ar (temp: disable openssl)

aharpervc committed:
f: forgot to turn on ridk first

bryanwieg commented:
I'm not an expert with some of this, but I was able to replicate this issue from in a local container finally. It's a base Windows Server container, and I just log in and follow the same process as set out in the circleci config.yml. (with some minor differences) 

I can get all the way to the point of the `bundle exec rake ports:libiconv`.

From inside the container, I see it gets stuck actually, here: 
```
Extracting libiconv-1.15.tar.gz into tmp/x86_64-w64-mingw32/ports/libiconv/1.15... OK
Running 'configure' for libiconv 1.15...
```
And further, there is a build log (is this right?), showing it get's stuck here:
C:/Users/ContainerAdministrator/repo/tiny_tds/tmp/x86_64-w64-mingw32/ports/libiconv/1.15/configure.log
```
checking minix/config.h presence... no
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking whether _XOPEN_SOURCE should be defined... no
checking for Minix Amsterdam compiler... no
checking for x86_64-w64-mingw32-ar...
```

After that last line, it sits. I even tried making sure that ` x86_64-w64-mingw32-ar` package is present, and from what I can tell it is.

There is no CPU usage, no activity, it's just stuck, waiting. This is my first time learning about these make processes. So today I stop at trying to find out why it is hung at this configure-> checking step. 

I am also somewhat curious as to why this is happing here, as this issue doesn't seem to show up anywhere else on google.

bryanwieg commented:
I'm just sharing this to see if you have any input...

I've learned that the repo uses miniportile2 to build these libs as part of the rake.

Well, miniportile2 is executing the `./configure`. If I put miniportile2 into debug, it will spit out the full command it is running.

```
DEBUG: env is {}
DEBUG: command is ["sh", "./configure", "--host=x86_64-w64-mingw32", "--disable-static", "--enable-shared", "CFLAGS=-fPIC -O2", "--prefix=C:/Users/ContainerAdministrator/repo/tiny_tds/ports/x86_64-w64-mingw32/libiconv/1.16"]
Running 'configure' for libiconv 1.16... DEBUG: options are {[:out, :err]=>["C:/Users/ContainerAdministrator/repo/tiny_tds/tmp/x86_64-w64-mingw32/ports/libiconv/1.16/configure.log", "a"]}
```

The crazy thing is that if I take that command and run in from `ridk exec`on the prompt directly - minus the quotes and commas - , it works! I can't figure out yet why it runs differently when miniportile2 does it.

```
sh ./configure --host=x86_64-w64-mingw32 --disable-static --enable-shared --prefix=C:\Users\ContainerAdministrator\repo\tiny_tds\tmp\x86_64-w64-mingw32\ports\libiconv\1.15\libiconv-1.15
```

Another observation, the openssl task builds it's own configure command, overriding the [miniportile2 method](https://github.com/flavorjones/mini_portile/blob/b9f0763088c460b6761d70a925e2861502891929/lib/mini_portile2/mini_portile.rb#L109):

./ports/openssl.rb
```
    def configure
      return if configured?

      md5_file = File.join(tmp_path, 'configure.md5')
      digest   = Digest::MD5.hexdigest(computed_options.to_s)
      File.open(md5_file, "w") { |f| f.write digest }

      # Windows doesn't recognize the shebang so always explicitly use sh
      execute('configure', "sh -c \"./Configure #{computed_options.join(' ')}\"")
    end
```

It runs this:
```
DEBUG: env is {}
DEBUG: command is "sh -c \"./Configure shared mingw64 --openssldir=C:/Users/ContainerAdministrator/repo/tiny_tds/ports/x86_64-w64-mingw32/openssl/1.1.1d --prefix=C:/Users/ContainerAdministrator/repo/tiny_tds/ports/x86_64-w64-mingw32/openssl/1.1.1d\""
Running 'configure' for openssl 1.1.1d... DEBUG: options are {[:out, :err]=>["C:/Users/ContainerAdministrator/repo/tiny_tds/tmp/x86_64-w64-mingw32/ports/openssl/1.1.1d/configure.log", "a"]}
OK
```

And openssl builds fine.

And as a matter of fact, if I remove the override and let miniportile2 build the command it's way, it spits out this, and openssl also fails to build. Albeit, with a different error during the configure process.

```
DEBUG: env is {}
DEBUG: command is ["sh", "./configure", "shared", "mingw64", "--openssldir=C:/Users/ContainerAdministrator/repo/tiny_tds/ports/x86_64-w64-mingw32/openssl/1.1.1d", "--prefix=C:/Users/ContainerAdministrator/repo/tiny_tds/ports/x86_64-w64-mingw32/openssl/1.1.1d"]
Running 'configure' for openssl 1.1.1d... DEBUG: options are {[:out, :err]=>["C:/Users/ContainerAdministrator/repo/tiny_tds/tmp/x86_64-w64-mingw32/ports/openssl/1.1.1d/configure.log", "a"]}
```

I don't know if it is a rabbit hole. No one reports any issues like this on the miniportile repo. So I assume it is something we are doing here. I can't figure out what is so different from how it was built in appveyor. 

Just seemed silly to try to work around miniportile2 with overrides, perhaps there is something else in the project that needs to be fixed.

In any case, I'm still cracking at it, and open to thoughts.

bryanwieg commented:
This is the docker image I am working from to avoid using up all the circleci minutes. You can see I switched back to choco from scoop, as you had earlier done. I spend a few days working on it with scoop, and found that choco vs scoop didn't seem to make a difference. It hasn't affected the actually problem. 

I only add vsbuild tools because that's what circleci says in in their image.

From here I just toy around with various things trying to get libiconv to build.

```
# syntax=docker/dockerfile:1
FROM mcr.microsoft.com/windows/server:ltsc2022

RUN powershell.exe -Command \
  $ErrorActionPreference = 'Stop'; \
  Set-ExecutionPolicy Bypass -Scope Process -Force; [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor 3072; iex ((New-Object System.Net.WebClient).DownloadString('https://community.chocolatey.org/install.ps1')) ; \
  choco install visualstudio2019buildtools -y ; \
  choco install git -y ; \
  choco install ruby --version 2.7.5.1 -y ; \
  choco install msys2 -y ; \
  choco install micro -y ;

RUN powershell.exe -Command \
  $ErrorActionPreference = 'Stop'; \
  ridk install 2 3 ; \
  mkdir ~\repo ; \
  cd ~\repo ; \
  git clone https://github.com/veracross/tiny_tds.git ; \
  cd .\tiny_tds ; \
  git checkout circleci ; \
  ridk exec bundle config set --local path './vendor/bundle' ; \
  ridk exec bundle install --jobs=3 --retry=3

ENTRYPOINT powershell
```

aharpervc commented:
> The crazy thing is that if I take that command and run in from `ridk exec`on the prompt directly - minus the quotes and commas - , it works! I can't figure out yet why it runs differently when miniportile2 does it.

This is great, I wasn't able to figure out how to get more debugging info. I wonder if that's actually enough to unblock this? Eg, instead of running `bundle exec rake ports:libiconv`, run `ridk exec sh ./configure ...` manually?

aharpervc committed:
wip: enable miniportile debugging

aharpervc committed:
wip: change scoop install for admins
- per their instructions
- I guess this has changed since the last time the job ran

aharpervc committed:
wip: disable extra packages for the moment

bryanwieg commented:
I imagine we still need the rake, as it does more than the configure- it also downloads the package and does the make. I suppose if it comes down to it we could just try overriding the miniportile2 `execute` as openssl does. I just wish adding more code, more override code, wasn't the solution. xD

I was going to spend more time trying to absorb what miniportile does. As in, how much do we need it? Are we using it the right way? That sort of thing.

aharpervc committed:
f: don't put the install script where we need to clone the repo
- duh

aharpervc commented:
Yeah, that's a good point. Also, why did we have to override `configure` for openssl, but apparently _not_ libiconv? The answer could be lost to the mists of time but it seems suspicious

e: maybe before/after this https://github.com/flavorjones/mini_portile/blob/6e3b08ce5ab768921f04ae0aa6cc96a149cfb6ff/lib/mini_portile2/mini_portile.rb#L136-L139

aharpervc committed:
libiconv configure experiment 1
- clone configure method from openssl

bryanwieg commented:
I tried forcing it to run with an older version of miniportile, 2.5.3 when those lines were last changed, and it still didn't work. :(

I agree, it felt like this miniportile was added years ago, but they had to jump through a few hoops to get it to work. I'm not experienced with what is happening here in the build process, but my first (uninformed) question was, can it just be removed in favor of something simpler.

aharpervc committed:
libiconv configure experiment 2
- try running ridk exec manually

aharpervc committed:
libiconv experiment 2.1
- absolute path to configure script? (might need to cd first, we'll see...)

aharpervc committed:
libiconv 2.2: try another path to configure

aharpervc committed:
f: duh

aharpervc commented:
Progress? [A new error?](https://app.circleci.com/pipelines/github/veracross/tiny_tds/32/workflows/e244cd73-5cee-4f27-be80-296770eea94e/jobs/38?invite=true#step-111-22)

> configure: WARNING: you should use --build, --host, --target
> configure: error: unrecognized option: `-fPIC'

aharpervc commented:
I wonder if the CFLAGS are supposed to be quoted, like `-fPIC -O2` together? 🤔

aharpervc committed:
try quoting CFLAGS value?

aharpervc commented:
Well that just gives me `error: unrecognized option: -fPIC -O2`, so not sure

aharpervc committed:
wip: try without cflags?

aharpervc commented:
Removing CLFAGS worked, so I assume there's probably a parameter quoting issue there. The next issue is that configure stalls out on detecting `ar`. It hangs on `checking for x86_64-w64-mingw32-ar...`

aharpervc committed:
wip: patch msys?
- https://github.com/msys2/msys2-runtime/issues/77#issuecomment-1079221137

aharpervc committed:
f: duh
- can only copy files from the repo after the repo is checked out

aharpervc committed:
f: path/file diagnostics

bryanwieg commented:
Hi Andrew,

I just wanted to let you know I have not forgotten about this. It's just a matter of allocating the available time. It is still on my todo list.

Bryan

bryanwieg commented:
After many hours and so much learning about the gcc build process--- I narrowed it down to differences in the build environment, not code in miniportile2 or the libs. rudydev and msys2 installation method and versions of each of those packages were critical. also the environment variables from appveyor.

So, in the end, installing via choco just didn't work. Installing ruby and msys2 separately didn't work. Instead, installing the rubydevkit (with the correct corresponding msys2 version included), using it's native installer- all this finally did the trick.

https://app.circleci.com/pipelines/github/bryanwieg/tiny_tds/12/workflows/7132cab6-feef-4f96-bb8c-109b6eb71910/jobs/12

https://github.com/rails-sqlserver/tiny_tds/pull/514/files

It seems like it is going to be a decent amount of work to rebuild this pipeline. But it's progress I suppose.

aharpervc commented:
Victory! ...sort of. This seems really strange, doesn't it? Should msys from rubyinstaller be "the same" as msys installed directly?

aharpervc committed:
wip: minus the msys patch, plus some env vars

aharpervc commented:
@bryanwieg regardless, should we build upon your PR to replace appveyor & travis with CircleCI? I'll post a comment over there

bryanwieg commented:
<details><summary>the long story</summary>
It's sort of a rabbit hole. Installing msys2 directly will install it's latest version. Which I did test using, but it would hang at building freetds. (Actually, the entire thing was a progression. After a change I would get to a hang at openssl, then change, then hang at libicov, then freetds... so much...)

In any case,

I noticed that the native ruby-devkit installer for 2.7.1 was using a version of msys2 from 2018-2019. However, msys2 at that time did not have an installer that would work non-interactively. It actually wasn't until sometime in 2020 that msys2 could install that way on it's own. And other stuff as well. 

So installing msys2 non-interactively.. especially via choco, would grab the latest or a much newer version than was in use at the time of ruby 2.7.1 development.

The only way to get the version of msys2 that was distributed at the time of ruby 2.7.1, and for it to install non-interactively, and hook together correctly, was to use the rubyinstall 2.7.1 devkit that had msys2 bundled inside. 

As for choco, there was something about the way it installed both packages where pathing and connecting to the ruby environment never worked quite right. 

It was amazing finally, that when I just used the bundled installer, everything just clicked.
</details>

Sure, I haven't made it to the testing and entire environment for that yet. But that seems like the next step, unless there is something I am missing.
